backend/src/interactions/updateUnconfirmedContribution/UnconfirmedContributionAdmin.role.ts

```diff
@@ -0,0 +1,56 @@
+import { Contribution } from '@entity/Contribution'
+import { User } from '@entity/User'
+
+import { RIGHTS } from '@/auth/RIGHTS'
+import { Role } from '@/auth/Role'
+import { AdminUpdateContributionArgs } from '@/graphql/arg/AdminUpdateContributionArgs'
+import { ContributionStatus } from '@/graphql/enum/ContributionStatus'
+import { LogError } from '@/server/LogError'
+
+import { UnconfirmedContributionRole } from './UnconfirmedContribution.role'
+
+export class UnconfirmedContributionAdminRole extends UnconfirmedContributionRole {
+  public constructor(
+    contribution: Contribution,
+    private updateData: AdminUpdateContributionArgs,
+    private moderator: User,
+  ) {
+    super(
+      contribution,
+      updateData.amount ?? contribution.amount,
+      updateData.creationDate ? new Date(updateData.creationDate) : contribution.contributionDate,
+    )
+  }
+
+  protected update(): void {
+    this.self.amount = this.updatedAmount
+    this.self.memo = this.updateData.memo ?? this.self.memo
+    this.self.contributionDate = this.updatedCreationDate
+    this.self.contributionStatus = ContributionStatus.PENDING
+    this.self.updatedAt = new Date()
+    this.self.updatedBy = this.moderator.id
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  protected checkAuthorization(user: User, role: Role): UnconfirmedContributionRole {
+    if (
+      !role.hasRight(RIGHTS.MODERATOR_UPDATE_CONTRIBUTION_MEMO) &&
+      this.self.moderatorId === null
+    ) {
+      throw new LogError('An admin is not allowed to update an user contribution')
+    }
+    return this
+  }
+
+  protected async validate(clientTimezoneOffset: number): Promise<void> {
+    await super.validate(clientTimezoneOffset)
+    // creation date is currently not changeable
+    if (
+      this.self.memo === this.updateData.memo &&
+      this.self.amount === this.updatedAmount &&
+      this.self.contributionDate.getTime() === new Date(this.updatedCreationDate).getTime()
+    ) {
+      throw new LogError("the contribution wasn't changed at all")
+    }
```

Comment:
I assume to get valid Data from admin. The admin interface should only trigger the resolver with valid data. 
So the norm is I  get valid data. If someone try to hack the API than maybe it gives invalid data and it is time for a exception. Currently I don't use try and catch at all, I let catch GraphQL the error and return it as error. 
If I return an expressive result maybe the caller don't even check it and wrote a invalid Entity into db.

Reply:
Ouh i see...
It was not clear for me that this method was directly invoked from our public API by the admin interface.
I thought there is an invocation chain with more than one invocations after the graphQL-Resolver in between before invoking the `validate` method.
It's ok for me if the invokation is currently nested in a try-catch-block, it was simply a question of efficient reusage in other invokation contexts.

Reply:
ContributionResolver.updateContribution => new UpdateUnconfirmedContributionContext() => run() => AbstractUnconfirmedContributionRole.checkAndUpdate() => validate

An Exception can easly travel up to the caller, expressive result would make much extra work in this case. 
In my first NodeServer C++ I used mainly expressive result until I found out the correct usage of exceptions. Than I could refactor my code, and make it much shorter and easier to read. Since then I decide carefully if I use exceptions or expressive results.